README.md
# healthgroup

[![GoTemplate](https://img.shields.io/badge/go/template-black?logo=go)](https://github.com/SchwarzIT/go-template)

`healthgroup` is a tool to check the health of services with an auxiliary group of health checks.

If one of the checks in a group fails, healthgroup returns the `503` status code.

Use cases where `healthgroup` can be helpful are up to your imagination :) The typical use case is a situation where you want to make an HTTP(S) Load Balancer aware of a service's health that is not placed directly in the backend, and between the LB and the service is a hop, e.g. proxy.

## Table of Contents

- [healthgroup](#healthgroup)
  - [Table of Contents](#table-of-contents)
  - [Usage](#usage)
  - [Example of usage](#example-of-usage)
  - [Configuration](#configuration)
    - [Environment variables](#environment-variables)
    - [Configuration file](#configuration-file)
      - [HTTP(s) health check specification](#https-health-check-specification)
    - [Read configuration file from Kubernetes ConfigMap](#read-configuration-file-from-kubernetes-configmap)
  - [Health check grouping](#health-check-grouping)
  - [Endpoints](#endpoints)
    - [Kubernetes](#kubernetes)
      - [Path Parameters](#path-parameters)
      - [Sample Request](#sample-request)
      - [Sample Response](#sample-response)
    - [Consul](#consul)
      - [Path Parameters](#path-parameters-1)
      - [Query Parameters](#query-parameters)
      - [Sample Request](#sample-request-1)
      - [Sample Response](#sample-response-1)
  - [Test \& lint](#test--lint)

## Usage

```text
Usage of healthgroup:
      --config string       config file (default is $HOME/.healthgroup.yaml)
      --configmap string    ConfigMap with the configuration file, e.g. namespace/configmap
      --in-cluster          use in-cluster config. Use always in a case when the app is running on a Kubernetes cluster
      --kubeconfig string   absolute path to the kubeconfig file (default "/Users/tczekajlo/.kube/config")
```

## Example of usage

```text
healthgroup --config ./test/testdata/config.yaml
[...]
{"level":"info","ts":1681571453.370432,"caller":"healthcheck/healthcheck.go:133","msg":"external health check","request_id":"131d2a2d-8e75-4926-8776-fe2e0470ce31","url":"https://google.com","status":200}
{"level":"info","ts":1681571453.3706791,"caller":"logger/logger.go:70","msg":"request","request_id":"131d2a2d-8e75-4926-8776-fe2e0470ce31","duration_ms":319,"status":200,"url":"/health/kubernetes/default/kubernetes","agent":"curl/7.86.0","ip":"127.0.0.1","method":"GET"}
```

## Configuration

In this section, you can learn how to configure `healthgroup`. The configuration is read in the following order:

- environment variables
- configuration file

The environment variables always take precedence over the configuration file.

### Environment variables

| Variable                         | Description                                                                                                                                            |
|----------------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------|
| `HG_CONSUL_TIMEOUT`              | Timeout specifies a time limit for requests made to the Consul server. A Timeout of zero means no timeout.                                             |
| `HG_CONSUL_ADDRESS`              | This is the address of the Consul server.                                                                                                              |
| `HG_CONSUL_SCHEME`               | This is the URI scheme for the Consul server.                                                                                                          |
| `HG_CONSUL_TOKEN`                | This is the API access token required when access control lists (ACLs) are enabled.                                                                    |
| `HG_CONSUL_INSECURE_SKIP_VERIFY` | This is a boolean value (default `true`) to specify SSL certificate verification; setting this value to `false` is not recommended for production use. |
| `HG_CONSUL_KEY_FILE`             | Path to a client key file to use for TLS.                                                                                                              |
| `HG_CONSUL_CERT_FILE`            | Path to a client cert file to use for TLS.                                                                                                             |
| `HG_CONSUL_CA_FILE`              | Path to a CA file to use for TLS when communicating with Consul.                                                                                       |
| `HG_LOG_LEVEL`                   | Defines log level.                                                                                                                                     |
| `HG_CONSUL_ENABLED`              | Defines if requests to Consul should be enabled.                                                                                                       |
| `HG_KUBERNETES_ENABLED`          | Defines if request to Kubernetes should be enabled.                                                                                                    |
| `HG_CONCURRENCY`                 | Defines how many health checks can be executed in parallel.                                                                                            |
| `HG_SERVER_PORT`                 | Defines a port on which to listen to.                                                                                                                  |
| `HG_SERVER_ADDRESS`              | Defines a bind address of the server.                                                                                                                  |

### Configuration file

In this section you can find a configuration file with default values. The configuration file is read from the `$HOME/.healthgroup.yaml` location by default. You can use the `--config` flag to define a path to the configuration file.

Under the [`configs/config.yaml`](./configs/config.yaml) path, you can find an example of the configuration file with all parameters.

```yaml
# example
server:
  address: 127.0.0.1
  port: 8080
concurrency: 5
kubernetes:
  enabled: true
httpHealthCheck:
  - timeout: 3s
    type: https
    host: google.com
```

| Parameter                   | Description                                                                                                                       | Type                | Default          |
|-----------------------------|-----------------------------------------------------------------------------------------------------------------------------------|---------------------|------------------|
| `server.port`               | Defines a port on which to listen to                                                                                              | `int`               | `8080`           |
| `server.idleTimeout`        | The maximum amount of time to wait for the next request (when keep-alive is enabled)                                              | `string`            | `5s`             |
| `server.address`            | Settings bind address                                                                                                             | `string`            | `0.0.0.0`        |
| `kubernetes.enabled`        | Defines if Kubernetes discovery service should be enabled                                                                         | `bool`              | `true`           |
| `httpHealthCheck`           | Defines auxiliary HTTP(S) health checks                                                                                           | `httpHealthCheck[]` | `[]`             |
| `consul.token`              | The API access token                                                                                                              | `string`            | `""`             |
| `consul.timeout`            | Timeout specifies a time limit for requests made to the Consul server. A Timeout of zero means no timeout, e.g. `2s`, `30s`, `1h` | `string`            | `2s`             |
| `consul.scheme`             | The URI scheme for the Consul server (available: `http` \| `https`)                                                               | `string`            | `http`           |
| `consul.keyFile`            | Path to a client key file to use for TLS                                                                                          | `string`            | `""`             |
| `consul.insecureSkipVerify` | Specify SSL certificate verification                                                                                              | `bool`              | `false`          |
| `consul.enabled`            | Defines if Consul discovery service should be enabled                                                                             | `bool`              | `false`          |
| `consul.certFile`           | Path to a client cert file to use for TLS                                                                                         | `string`            | `""`             |
| `consul.caFile`             | Path to a CA file to use for TLS when communicating with Consul                                                                   | `string`            | `""`             |
| `consul.address`            | The address of the Consul server                                                                                                  | `string`            | `127.0.0.1:8500` |
| `concurrency`               | Defines how many health checks can be executed in parallel per request                                                            | `int`               | `5`              |

#### HTTP(s) health check specification

| Health check parameter | Description                                                                                               | Type     | Default |
|------------------------|-----------------------------------------------------------------------------------------------------------|----------|---------|
| `host`                 | Address of the target to which the probe should connect                                                   | `string` | `""`    |
| `insecureSkipVerify`   | Whether to verify SSL certificate                                                                         | `bool`   | `false` |
| `namespace`            | The namespace name that the check should be group with                                                    | `string` | `""`    |
| `port`                 | Port of the target host                                                                                   | `int`    | `null`  |
| `requestPath`          | The request path to which the probe should connect                                                        | `string` | `/`     |
| `service`              | The service name that the check should be group with                                                      | `string` | `""`    |
| `timeout`              | Timeout specifies a time limit for requests made to the Consul server. A Timeout of zero means no timeout | `string` | `0s`    |
| `type`                 | Type of the check, available: `http`, `https`, `http2`                                                    | `string` | `http`  |
| `discovery`            | Specifies a discovery service for which the check should be executed, available: `consul`, `kubernetes`   | `string` | `""`    |

### Read configuration file from Kubernetes ConfigMap

It's possible to read the configuration file directly from a Kubernetes ConfigMap. The [`configs/configmap-healthgroup.yaml`](./configs/configmap-healthgroup.yaml) file shows an example of ConfigMap that includes the configuration file for healgroup. The configuration file has to be passed under the `config.yaml` key.

Using the `--configmap` flag you can define which ConfigMap should be used. The flag value has the following syntax: `namespace/configmap_name`.

## Health check grouping

By default, all auxiliary checks are executed along with the main one (every time when the `/health/*` endpoint is called). If you want to assign a particular check to a given service, namespace, or discovery, you can do it by defining the `service` or/and `namespace`, or/and `discovery` parameters in the health check specification.

For instance, if you'd like to check health of the `https://google.com` every time when you check the health of the `mytest` Kubernetes service that is located in the `staging` namespace. It's what the configuration would look like:

```yaml
httpHealthCheck:
  - type: https
    host: google.com
    namespace: staging
    service: mytest
    discovery: kubernetes
```

If one of the grouping parameters is omitted, it works like a wildcard.

```yaml
httpHealthCheck:
  # Execute the health check if the service is a Kubernetes service,
  # located in the staging namespace, and the name of the service is `mytest`.
  - type: https
    host: google.com
    namespace: staging
    service: mytest
    discovery: kubernetes
  # Execute the health check if the service name is `mytest`.
  - type: https
    host: example.com
    service: mytest
  # Execute the health check every time.
  - type: https
    host: github.com
```

## Endpoints

Below you can find a list of endpoints supported by `healthgroup`.

### Kubernetes

The Kubernetes endpoint checks the status of a Kubernetes service. If the Kubernetes services don't have any healthy endpoints then the endpoint returns the `503` status code.

If any of the auxiliary health checks failed, the endpoint returns the `503` status code.

| Method | Path                                     | Produces           |
|--------|------------------------------------------|--------------------|
| `GET`  | `/health/kubernetes/:namespace/:service` | `application/json` |

#### Path Parameters

- `namespace` `(string: <required>)` - Specifies the name of the namespace where the Kubernetes service is located.
- `service` `(string: <required>` - Specifies the name of the Kubernetes service.

#### Sample Request

```bash
curl -i http://localhost:8080/health/kubernetes/default/kubernetes
```

#### Sample Response

```text
HTTP/1.1 200 OK
Content-Type: application/json
Content-Length: 53
X-Request-Id: 88ab63e5-5cbb-49eb-9840-e855e3fab1d6
```

```json
{
  "success": true,
  "message": "all health checks passed"
}
```

### Consul

The Consul endpoint checks the status of a Consul service. If the Consul services return zero healthy instances then the endpoint returns the `503` status code.

If any of the auxiliary health checks failed, the endpoint returns the `503` status code.

| Method | Path | Produces |
| -- | --| -- |
| `GET` | `/health/consul/:service` | `application/json` |
| `GET` | `/health/consul/:namespace/:service` | `application/json` |

#### Path Parameters

- `namespace` `(string: <optional>)` - Specifies the name of the namespace where the Consul service is located. The parameter works only with Consul Enterprise.
- `service` `(string: <required>)` - Specifies the name of the Consul service.

#### Query Parameters

- `tag` `(string: "")` - Specifies the tag to filter the list of instances for a given service.

#### Sample Request

```bash
curl -i http://127.0.0.1:8080/health/consul/redis
```

#### Sample Response

```text
HTTP/1.1 503 Service Unavailable
Content-Type: application/json
Content-Length: 52
X-Request-Id: 84ab63e5-5cbb-49eb-9840-e855e3fab1d6
```

```json
{
  "success": false,
  "message": "Service is not healthy"
}
```

## Test & lint

Run linting

```bash
make lint
```

Run tests

```bash
make test
```

Whenever you need help regarding the available actions, use the `make help` command.


CHANGELOG.md

<a name="1.1.1"></a>
## [1.1.1](https://github.com/tczekajlo/healthgroup/compare/1.1.0...1.1.1) (2023-06-24)

### Bug Fixes

* bump golang.org/x/net from 0.10.0 to 0.11.0 ([#25](https://github.com/tczekajlo/healthgroup/issues/25))
* bump k8s.io/client-go from 0.27.2 to 0.27.3 ([#23](https://github.com/tczekajlo/healthgroup/issues/23))
* bump github.com/gofiber/fiber/v2 from 2.46.0 to 2.47.0 ([#26](https://github.com/tczekajlo/healthgroup/issues/26))
* bump github.com/spf13/viper from 1.15.0 to 1.16.0 ([#18](https://github.com/tczekajlo/healthgroup/issues/18))
* bump github.com/hashicorp/consul/api from 1.20.0 to 1.21.0 ([#19](https://github.com/tczekajlo/healthgroup/issues/19))
* bump github.com/stretchr/testify from 1.8.3 to 1.8.4 ([#20](https://github.com/tczekajlo/healthgroup/issues/20))

### Chore

* Disable depguard for golangci-lint ([#21](https://github.com/tczekajlo/healthgroup/issues/21))


<a name="1.1.0"></a>
## [1.1.0](https://github.com/tczekajlo/healthgroup/compare/1.0.3...1.1.0) (2023-05-28)

### Chore

* Update the release changelog template

### Features

* Read configuration file from Kubernetes ConfigMap ([#17](https://github.com/tczekajlo/healthgroup/issues/17))


<a name="1.0.3"></a>
## [1.0.3](https://github.com/tczekajlo/healthgroup/compare/1.0.2...1.0.3) (2023-05-27)

### Bug Fixes

* bump github.com/gofiber/fiber/v2 from 2.45.0 to 2.46.0 ([#14](https://github.com/tczekajlo/healthgroup/issues/14))
* bump github.com/stretchr/testify from 1.8.2 to 1.8.3
* bump k8s.io/client-go from 0.27.1 to 0.27.2


<a name="1.0.2"></a>
## [1.0.2](https://github.com/tczekajlo/healthgroup/compare/1.0.1...1.0.2) (2023-05-09)

### Bug Fixes

* bump golang.org/x/sync from 0.1.0 to 0.2.0
* bump golang.org/x/net from 0.9.0 to 0.10.0
* bump github.com/gofiber/fiber/v2 from 2.44.0 to 2.45.0

### Chore

* update changelog for 1.0.2


<a name="1.0.1"></a>
## [1.0.1](https://github.com/tczekajlo/healthgroup/compare/1.0.0...1.0.1) (2023-04-16)

### Bug Fixes

* Update dependencies
* Bump golang.org/x/net from 0.8.0 to 0.9.0 ([#7](https://github.com/tczekajlo/healthgroup/issues/7))
* Bump github.com/gofiber/fiber/v2 from 2.43.0 to 2.44.0 ([#3](https://github.com/tczekajlo/healthgroup/issues/3))

### Chore

* Show git sha in the release changelog


<a name="1.0.0"></a>
## [1.0.0](https://github.com/tczekajlo/healthgroup/compare/0.2.0...1.0.0) (2023-04-16)

### Chore

* update changelog for 1.0.0
* Add labels to the container image


<a name="0.2.0"></a>
## 0.2.0 (2023-04-16)



internal/config/config.go
package config

import (
	"bytes"
	"context"
	"fmt"
	"os"
	"strings"
	"time"

	"github.com/mitchellh/go-homedir"
	"github.com/spf13/viper"
	"go.uber.org/zap"
	"golang.org/x/xerrors"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/client-go/kubernetes"
	"k8s.io/client-go/rest"
	"k8s.io/client-go/tools/clientcmd"
)

func New(opts ...Option) *Config {
	config := &Config{}

	for _, opt := range opts {
		opt(config)
	}

	return config
}

// ReadConfigFromFile reads the configuration from a Kubernetes ConfigMap.
func (c *Config) ReadConfigFromConfigMap() error {
	var (
		config *rest.Config
		err    error
	)

	flags := c.Flags()

	if flags.InCluster {
		config, err = rest.InClusterConfig()
		if err != nil {
			return err
		}
	} else {
		config, err = clientcmd.BuildConfigFromFlags("", c.flags.Kubeconfig)
		if err != nil {
			return err
		}
	}
	// creates the clientset
	httpClient, err := rest.HTTPClientFor(config)
	if err != nil {
		return err
	}

	clientset, err := kubernetes.NewForConfigAndClient(config, httpClient)
	if err != nil {
		return err
	}

	cfg := strings.Split(c.flags.ConfigMap, "/")

	if len(cfg) != 2 { //nolint:gomnd
		return xerrors.Errorf("ConfigMap doesn't exists: %s", cfg)
	}

	ns := cfg[0]
	name := cfg[1]

	c.logger.Info("Reading configuration from ConfigMap",
		zap.String("namespace", ns),
		zap.String("name", name),
	)
	cm, err := clientset.CoreV1().ConfigMaps(ns).Get(context.TODO(), name, metav1.GetOptions{})
	if err != nil {
		return err
	}

	r := bytes.NewBufferString(cm.Data["config.yaml"])
	viper.SetConfigType("yaml")

	return viper.ReadConfig(r)
}

// ReadConfigFromFile reads the configuration from a file.
func (c *Config) ReadConfigFromFile() {
	if c.file != "" {
		// Use config file from the flag.
		viper.SetConfigFile(c.file)
	} else {
		// Find home directory.
		home, err := homedir.Dir()
		if err != nil {
			fmt.Println(err)
			os.Exit(1)
		}

		// Search config in home directory with name ".healthgroup" (without extension).
		viper.AddConfigPath(home)
		viper.SetConfigName(".healtgroup")
		viper.SetConfigType("yaml")
	}

	// If a config file is found, read it in.
	if err := viper.ReadInConfig(); err == nil {
		c.logger.Info("Using config", zap.String("file", viper.ConfigFileUsed()))
	}
}

func (c *Config) ReadConfig() error {
	if err := c.SetDefault(); err != nil {
		return err
	}

	if c.flags.ConfigMap != "" {
		if err := c.ReadConfigFromConfigMap(); err != nil {
			return err
		}
	} else {
		c.ReadConfigFromFile()
	}

	err := viper.Unmarshal(&c)
	if err != nil {
		panic("configuration file: unable to decode into struct")
	}

	if err := c.SetFromEnv(); err != nil {
		return err
	}

	c.logger.Info("Configuration", zap.Any("config", c))

	return nil
}

func (c *Config) SetDefault() error {
	c.logger.Debug("setting default values")

	c.Server.Address = "0.0.0.0"
	c.Server.Port = 8080
	c.Server.IdleTimeout = time.Second * 5 //nolint:gomnd
	c.Concurrency = 5
	c.Kubernetes.Enabled = true
	c.Consul.Enabled = false
	c.Consul.Address = "127.0.0.1:8500"
	c.Consul.Scheme = "http"
	c.Consul.InsecureSkipVerify = false
	c.Consul.Timeout = time.Second * 2 //nolint:gomnd

	return nil
}

func (c *Config) SetFromEnv() error { //nolint:cyclop
	c.logger.Debug("reading environment variables and setting values")

	if v := viper.GetString("server_address"); v != "" {
		c.Server.Address = v
	}

	if v := viper.GetInt("server_port"); v != 0 {
		c.Server.Port = v
	}

	if v := viper.GetInt("concurrency"); v != 0 {
		c.Concurrency = v
	}

	_, ok := os.LookupEnv("HG_KUBERNETES_ENABLED")
	if v := viper.GetBool("kubernetes_enabled"); ok {
		c.Kubernetes.Enabled = v
	}

	_, ok = os.LookupEnv("HG_CONSUL_ENABLED")
	if v := viper.GetBool("consul_enabled"); ok {
		c.Consul.Enabled = v
	}

	if v := viper.GetString("consul_address"); v != "" {
		c.Consul.Address = v
	}

	if v := viper.GetString("consul_scheme"); v != "" {
		c.Consul.Scheme = v
	}

	if v := viper.GetString("consul_token"); v != "" {
		c.Consul.Token = v
	}

	if v := viper.GetString("consul_ca_file"); v != "" {
		c.Consul.CAFile = v
	}

	if v := viper.GetString("consul_cert_file"); v != "" {
		c.Consul.CertFile = v
	}

	if v := viper.GetString("consul_key_file"); v != "" {
		c.Consul.KeyFile = v
	}

	if v := viper.GetDuration("consul_timeout"); v != 0 {
		c.Consul.Timeout = v
	}

	_, ok = os.LookupEnv("HG_CONSUL_INSECURE_SKIP_VERIFY")
	if v := viper.GetBool("consul_insecure_skip_verify"); ok {
		c.Consul.InsecureSkipVerify = v
	}

	return nil
}

func (c *Config) Flags() *Flags {
	return c.flags
}
